cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala

```diff
@@ -0,0 +1,127 @@
+  numCores) {
+
+  def this(conf: SparkConf, numCores: Int) = {
+    this( conf,
+      SnappyUnifiedMemoryManager.getMaxMemory(conf),
+      numCores)
+  }
+  private val listener = Misc.getMemStoreBooting.thresholdListener()
+
+  private def freeMemory(blockId: Option[BlockId], numBytes: Long,
+      evictedBlocks: mutable.Buffer[(BlockId, BlockStatus)]): Boolean = {
+    val freeMemory = Runtime.getRuntime.freeMemory()
+    if (freeMemory < numBytes) {
+
+      storageMemoryPool.memoryStore.evictBlocksToFreeSpace(blockId,
+        numBytes - freeMemory,
+        evictedBlocks) || !listener.isCritical
+    }
+    false
+  }
+
+  override private[memory] def acquireExecutionMemory(
+      numBytes: Long,
+      taskAttemptId: Long,
+      memoryMode: MemoryMode): Long = synchronized {
+    assert(numBytes >= 0)
+    if (listener.isEviction || listener.isCritical) {
+      val evictedBlocks = new ArrayBuffer[(BlockId, BlockStatus)]
+      // If free memory not available, let Spark
+      // take a corrective action
+      if (!freeMemory(None, numBytes, evictedBlocks)) {
+        return 0
+      }
+    }
+    super.acquireExecutionMemory(numBytes, taskAttemptId, memoryMode)
+  }
+
+  override def acquireStorageMemory(
+      blockId: BlockId,
+      numBytes: Long,
+      evictedBlocks: mutable.Buffer[(BlockId, BlockStatus)]): Boolean = synchronized {
+    if (listener.isEviction || listener.isCritical) {
+      // If free memory not available, let Spark
+      // take a corrective action
+      if (!freeMemory(Some(blockId), numBytes, evictedBlocks)){
+        return false
+      }
+    }
+    super.acquireStorageMemory(blockId, numBytes, evictedBlocks)
+  }
+}
+
+private object SnappyUnifiedMemoryManager {
+
+  private val RESERVED_SYSTEM_MEMORY_BYTES = 300 * 1024 * 1024
+
+  /**
+   * Return the total amount of memory shared between execution and storage, in bytes.
+   * This is a direct copy from UnifiedMemorymanager with an extra check for evit fraction
+   */
+  private def getMaxMemory(conf: SparkConf): Long = {
+    val systemMemory = conf.getLong("spark.testing.memory", Runtime.getRuntime.maxMemory)
+    val reservedMemory = conf.getLong("spark.testing.reservedMemory",
+      if (conf.contains("spark.testing")) 0 else RESERVED_SYSTEM_MEMORY_BYTES)
+    val minSystemMemory = reservedMemory * 1.5
+    if (systemMemory < minSystemMemory) {
+      throw new IllegalArgumentException(s"System memory $systemMemory must " +
+          s"be at least $minSystemMemory. Please use a larger heap size.")
+    }
+    val usableMemory = systemMemory - reservedMemory
+    // GemFireXD is required to be already booted before constructing
+    // snappy memory manager
+    var evictFraction: Double = GemFireCacheImpl.getExisting.getResourceManager
+        .getEvictionHeapPercentage
+    if (evictFraction <= 0.0) {
+      evictFraction = 0.75
+    }
+    val memoryFraction = conf.getDouble("spark.memory.fraction", evictFraction)
```

Comment:
see comment above that this looks inconsistent with freeMemory that only fails after criticalFraction